jobs/gtfs-rt-parser/gtfs_rt_parser.py

```diff
@@ -123,109 +218,440 @@ def identify_files(glob, rt_file_type: RTFileType, progress=False) -> List[RTFil
-        if pbar:
-            pbar.update(1)
+    if written:
+        log(
+            f"writing {written} lines to {hour.data_hive_path}{JSONL_GZIP_EXTENSION}",
+            pbar=pbar,
+        )
+        put_with_retry(fs, gzip_fname, f"{hour.data_hive_path}{JSONL_GZIP_EXTENSION}")
+    else:
+        log(
+            f"WARNING: no records at all for {hour.data_hive_path}",
+            fg=typer.colors.YELLOW,
+            pbar=pbar,
+        )
 
+    if errors:
+        upload_gzipped_records(
+            fs,
+            gzip_fname=str(tmp_dir + f"/errors_{hour.suffix}" + JSONL_GZIP_EXTENSION),
+            out_path=f"{hour.errors_hive_path}{JSONL_GZIP_EXTENSION}",
+            records=errors,
+            pbar=pbar,
+            verbose=verbose,
+        )
+    else:
+        log(
+            f"WARNING: no records at all for {hour.errors_hive_path}",
+            fg=typer.colors.YELLOW,
+            pbar=pbar,
+        )
 
-def try_handle_one_feed(*args, **kwargs) -> Optional[Exception]:
-    try:
-        parse_file(*args, **kwargs)
-    except Exception as e:
-        typer.echo(
-            f"got exception while handling feed: {str(e)} from {str(e.__cause__ or e.__context__)} {traceback.format_exc()}",
-            err=True,
+
+# Originally this whole function was retried, but tmpdir flakiness will throw
+# exceptions in backoff's context, which ruins things
+def parse_and_aggregate_hour(
+    hour: RTAggregation,
+    jar_path: Path,
+    tmp_dir: str,
+    verbose: bool = False,
+    parse: bool = False,
+    validate: bool = False,
+    pbar=None,
+):
+    if not parse and not validate:
+        raise ValueError("skipping both parsing and validation does nothing for us!")
+
+    fs = get_fs()
+    suffix = hour.suffix
+    dst_path_gtfs = f"{tmp_dir}/gtfs_{suffix}/"
+    os.mkdir(dst_path_gtfs)
+    dst_path_rt = f"{tmp_dir}/rt_{suffix}/"
+    get_with_retry(
+        fs,
+        rpath=[file.path for file in hour.source_files],
+        lpath=[
+            os.path.join(dst_path_rt, file.timestamped_filename)
+            for file in hour.source_files
+        ],
+    )
+
+    if validate:
+        validate_and_upload(
+            fs=fs,
+            jar_path=jar_path,
+            dst_path_gtfs=dst_path_gtfs,
+            dst_path_rt=dst_path_rt,
+            tmp_dir=tmp_dir,
+            hour=hour,
+            verbose=verbose,
+            pbar=pbar,
+        )
+
+    if parse:
+        parse_and_upload(
+            fs=fs,
+            dst_path_rt=dst_path_rt,
+            tmp_dir=tmp_dir,
+            hour=hour,
+            verbose=verbose,
+            pbar=pbar,
         )
-        # a lot of these are thrown by tmpdir, and are potentially flakes; but if they were thrown during
-        # handling of another exception, we want that exception still
-        if isinstance(e, OSError) and e.errno == 39:
-            return e.__cause__ or e.__context__
-        return e
 
 
 def main(
```

Comment:
**super nitpick / very much just my opinion**: seems like kind of a lot going on for a `main` -- is there anything we can/should refactor into separate functions at all?

Reply:
Eh I think it's marginal, I don't necessarily like functions that are only called once for the sake of shortening a function.